src/coreclr/vm/interpframeallocator.cpp

```diff
@@ -0,0 +1,135 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#ifdef FEATURE_INTERPRETER
+
+#include "interpexec.h"
+#include "interpframeallocator.h"
+
+FrameDataAllocator::FrameDataFragment::FrameDataFragment(size_t size)
+{
+    // Amortize allocation cost by allocating a larger chunk of memory
+    if (size < INTERP_STACK_FRAGMENT_SIZE)
+    {
+        size = INTERP_STACK_FRAGMENT_SIZE;
+    }
+
+    pFrameStart = (uint8_t*)malloc(size);
+    if (pFrameStart != nullptr)
+    {
+        pFrameEnd = pFrameStart + size;
+        pFramePos = pFrameStart;
+    }
+    pNext = nullptr;
+}
+
+FrameDataAllocator::FrameDataFragment::~FrameDataFragment()
+{
+    free(pFrameStart);
+}
+
+FrameDataAllocator::FrameDataAllocator(size_t size)
+{
+    pFirst = new FrameDataFragment(size);
+    pCurrent = pFirst;
+    pInfos = nullptr;
+    infosLen = 0;
+    infosCapacity = 0;
+}
+
+FrameDataAllocator::~FrameDataAllocator()
+{
+    assert(pCurrent == pFirst && pCurrent->pFramePos == pCurrent->pFrameStart);
+    FreeFragments(pFirst);
+    free(pInfos);
+}
+
+void FrameDataAllocator::FreeFragments(FrameDataFragment *pFrag)
+{
+    while (pFrag)
+    {
+        FrameDataFragment *pNext = pFrag->pNext;
+        delete pFrag;
+        pFrag = pNext;
+    }
+}
+
+bool FrameDataAllocator::PushInfo(InterpMethodContextFrame *pFrame)
+{
+    if (infosLen == infosCapacity)
+    {
+        size_t newCapacity = infosCapacity == 0 ? 8 : infosCapacity * 2;
+        FrameDataInfo* newInfos = (FrameDataInfo*)realloc(pInfos, newCapacity * sizeof(FrameDataInfo));
+        if (newInfos == nullptr)
+        {
+            return false;
+        }
+        pInfos = newInfos;
+        infosCapacity = newCapacity;
+    }
+
+    FrameDataInfo *pInfo = &pInfos[infosLen++];
+    pInfo->pFrame = pFrame;
+    pInfo->pFrag = pCurrent;
+    pInfo->pFramePos = pCurrent->pFramePos;
+    return true;
+}
+
+void *FrameDataAllocator::Alloc(InterpMethodContextFrame *pFrame, size_t size)
+{
```

Comment:
Pointer returned by stackalloc is expected to be sufficiently aligned. For example, if somebody does `stackalloc byte[1]` followed by `stackalloc long[1]`, the second stackalloc needs to return pointer that is 8-byte aligned. The current implementation does appear to do that. Aligning the size up to multiple of pointer size is the easiest way to do that (CoreCLR has ALIGN_UP macro for it).

Reply:
It is aligned before the call: https://github.com/dotnet/runtime/pull/114860/files/4b8c03c5d29d4d9230072946b59f3fa5b5d18742

Let me know if you want it within the Alloc function.

Reply:
Ah ok. I think it would be better to have it in the Alloc function since it is required due to Alloc function implementation details. For example, if Alloc function called malloc for every stackallocs without coalescing them together, the alignment would not be needed.